src/core/atomic.d

```diff
@@ -32,12 +32,62 @@ else
 
 private
 {
-    template HeadUnshared(T)
+    /* Construct a type with a shared tail, and if possible with an unshared
+    head. */
+    template TailShared(T)
     {
-        static if( is( T U : shared(U*) ) )
-            alias shared(U)* HeadUnshared;
+        alias S = shared T;
+
+        static if (is(S U == shared U)) {}
+        else static assert(false);
```

Comment:
> Needs an error message

Added a message: "S should be shared." But I'm not sure if it's any good. The assert shouldn't ever be hit, no matter what `T` is. So I'm not sure what expectation to express in the message.